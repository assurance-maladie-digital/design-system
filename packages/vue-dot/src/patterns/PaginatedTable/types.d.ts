// Vuetify VDataTable options
export interface Options {
	page: number;
	itemsPerPage: number;
	sortBy: string[];
	sortDesc: boolean[];
	groupBy: string[];
	groupDesc: boolean[];
	multiSort: boolean;
	mustSort: boolean;
}
