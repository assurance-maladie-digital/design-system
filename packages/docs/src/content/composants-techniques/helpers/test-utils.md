---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
