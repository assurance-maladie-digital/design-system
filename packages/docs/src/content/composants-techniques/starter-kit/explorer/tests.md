Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
