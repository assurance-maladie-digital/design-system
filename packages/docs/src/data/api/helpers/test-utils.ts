import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
