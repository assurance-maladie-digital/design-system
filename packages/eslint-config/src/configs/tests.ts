import { type FlatESLintConfigItem } from 'eslint-define-config';
import { pluginVitest } from '../plugins';
import { GLOB_TESTS } from '../globs';
import { isInEditor } from '../env';

export const tests: FlatESLintConfigItem[] = [
	{
		plugins: {
			tests: pluginVitest,
		},
	},
	{
		files: GLOB_TESTS,
		rules: {
			'tests/consistent-test-it': [
				'error',
				{
					fn: 'it',
					withinDescribe: 'it',
				},
			],
			'tests/expect-expect': 'error',
			'tests/no-alias-methods': 'error',
			'tests/no-commented-out-tests': isInEditor ? 'off' : 'error',
			'tests/no-conditional-expect': 'error',
			'tests/no-disabled-tests': 'error',
			'tests/no-done-callback': 'error',
			'tests/no-duplicate-hooks': 'error',
			'tests/no-focused-tests': isInEditor ? 'off' : 'error',
			'tests/no-identical-title': 'error',
			'tests/no-interpolation-in-snapshots': 'error',
			'tests/no-standalone-expect': 'error',
			'tests/no-test-return-statement': 'error',
			'tests/prefer-comparison-matcher': 'error',
			'tests/prefer-equality-matcher': 'error',
			'tests/prefer-hooks-in-order': 'error',
			'tests/prefer-hooks-on-top': 'error',
			'tests/prefer-mock-promise-shorthand': 'error',
			'tests/prefer-spy-on': 'error',
			'tests/prefer-to-be-falsy': 'error',
			'tests/prefer-to-be-object': 'error',
			'tests/prefer-to-be-truthy': 'error',
			'tests/prefer-to-be': 'error',
			'tests/prefer-to-contain': 'error',
			'tests/prefer-to-have-length': 'error',
			'tests/prefer-todo': 'error',
			'tests/require-hook': 'error',
			'tests/require-top-level-describe': 'error',
			'tests/valid-describe-callback': 'error',
			'tests/valid-expect': 'error',
			'tests/valid-title': 'error',
		},
	},
];
