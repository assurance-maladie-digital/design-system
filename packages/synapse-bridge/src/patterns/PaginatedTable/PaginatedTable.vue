<script lang="ts">
import { defineComponent } from "vue";
import type { PropType } from "vue";
import type { DataOptions, SortOption, GroupOption } from "./types";
import { LocalStorageUtility } from "@/helpers/localStorageUtility";

export default defineComponent({
	inheritAttrs: false,
	emits: ['update:options'],
	props: {
		options: {
			type: Object as PropType<Partial<DataOptions>>,
			required: true,
		},
		serverItemsLength: {
			type: Number,
		},
		suffix: {
			type: String,
			default: undefined,
		},
		itemsPerPage: {
			type: Number
		},
	},
	data() {
		return {
			localStorageUtility: this.newLocalStorageInstance() as LocalStorageUtility,
			localOptions: {},
			slotNames: Object.keys(this.$slots)  as "default"[]
		};
	},
	watch: {
		options: {
			deep: true,
			handler() {
				if(this.serverItemsLength !== 0) {
					this.localStorageUtility.setItem(this.storageKey, {
						...this.optionsFacade,
						itemsLength: this.serverItemsLength ?? 0,
					});

					this.localOptions = this.optionsFacade;
				}
			},
		},
	},
	computed: {
		storageKey(): string {
			const prefix = 'pagination';

			return this.suffix ? `${prefix}-${this.suffix}` : prefix
		},
		headers(): undefined | { title?: string, value: string }[] {
			if (!Array.isArray(this.$attrs['headers'])) {
				return undefined;
			}
			return this.$attrs['headers'].map((header) => ({
				...header,
				title: header.title ?? header.text,
			}))
		},
		optionsFacade() {
			const sortBy =
				!this.options.sortBy ?
					[] :
					Array.isArray(this.options.sortBy) ?
						this.options.sortBy
							.filter((key: string) => key)
							.map((key: string, index: number) => (
								{ key, order: this.options?.sortDesc?.[index] ? 'desc' : 'asc' as 'desc' | 'asc'}
							)) :
						[{
							key: this.options.sortBy,
							order: this.options.sortDesc ? 'desc' : 'asc' as 'desc' | 'asc',
						}];
			const groupBy =
				!this.options.groupBy ?
					[] :
					Array.isArray(this.options.groupBy) ?
						this.options.groupBy
							.filter((key: string) => key)
							.map((key: string, index: number) => (
								{key, order: this.options?.groupDesc?.[index] ? 'desc' : 'asc'  as 'desc' | 'asc'}
							)) :
						[{
							key: this.options.groupBy,
							order: this.options.groupDesc ? 'desc' : 'asc' as 'desc' | 'asc',
						}];

			return {
				page: this.options.page || this.$attrs['page'],
				itemsPerPage: this.options.itemsPerPage || this.itemsPerPage,
				sortBy,
				groupBy,
				multiSort: this.options.multiSort,
				mustSort: this.options.mustSort,
			};
		},
		propsFacade() : any {
			const {
				"onUpdate:options": _,
				...attrs
			}: Record<string, any> = this.$attrs;

			return {
				...attrs,
				itemsLength: this.serverItemsLength ?? 0,
				headers: this.headers,
				...this.localOptions,
			};
		},
	},
	methods: {
		newLocalStorageInstance(): LocalStorageUtility {
			return new LocalStorageUtility();
		},
		updateOptions(options: SortOption[] | GroupOption[]): void {
			this.$emit('update:options', this.createUpdatedOptions(options));
		},
		createUpdatedOptions(options: any): DataOptions {
			return {
				...options,
				multiSort: this.optionsFacade.multiSort,
				mustSort: this.optionsFacade.mustSort,
				sortBy: this.createSortBy(options),
				sortDesc: this.createSortDesc(options),
				groupBy: this.createGroupBy(options),
				groupDesc: this.createGroupDesc(options),
			};
		},
		createSortBy(options: any): string[] {
			return options.sortBy
				.filter((sort: any) => sort.key)
				.map((sort: any) => sort.key);
		},
		createSortDesc(options: any): boolean[] {
			return options.sortBy
				.filter((sort: any) => sort.key)
				.map((sort: any) => sort.order === 'desc');
		},
		createGroupBy(options: any): string[] {
			return options.groupBy
				.filter((group: any) => group.key)
				.map((group: any) => group?.key);
		},
		createGroupDesc(options: any): boolean[] {
			return options.groupBy
				.filter((group: any) => group.key)
				.map((group: any) => group?.order === 'desc');
		}
	},
	created() {
		this.localOptions =
			this.localStorageUtility.getItem(this.storageKey) ?? this.optionsFacade;
	},
});
</script>

<template>
	<div>
		<VDataTable
			v-if="!serverItemsLength"
			color="warning"
			v-bind="propsFacade"
			@update:options="updateOptions"
		>
			<template
				v-for="(slotName) in slotNames"
				v-slot:[slotName]="slotProps"
			>
				<slot :name="slotName" v-bind="slotProps ?? {}" />
			</template>
		</VDataTable>
		<VDataTableServer
			v-else
			v-bind="propsFacade"
			@update:options="updateOptions"
		>
			<template
				v-for="(slotName) in slotNames"
				v-slot:[slotName]="slotProps"
			>
				<slot :name="slotName" v-bind="slotProps ?? {}" />
			</template>
		</VDataTableServer>
	</div>
</template>
