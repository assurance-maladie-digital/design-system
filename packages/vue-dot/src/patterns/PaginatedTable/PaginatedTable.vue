<template>
	<VDataTable
		v-if="$attrs"
		v-bind="$attrs"
		:options.sync="optionsCalc"
		v-on="$listeners"
	>
		<!--
			Pass down the scoped slots
			@see https://github.com/vuejs/vue/pull/7765 for the official discussion
			and https://stackoverflow.com/questions/50891858/vue-how-to-pass-down-slots-inside-wrapper-component/52823029#52823029
			for the source of this solution
		-->
		<template
			v-for="slot in Object.keys($scopedSlots)"
			#[slot]="scope"
		>
			<slot
				:name="slot"
				v-bind="scope"
			/>
		</template>

		<slot
			v-for="slot in Object.keys($slots)"
			:slot="slot"
			:name="slot"
		/>
	</VDataTable>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { DataOptions } from 'vuetify/types';

	import { LocalStorageUtility } from '../../helpers/localStorageUtility';

	const Props = Vue.extend({
		props: {
			options: {
				type: Object as PropType<DataOptions>,
				required: true
			},
			// The suffix is used to store different pagination objects
			// If the user has two tables but isn't using this prop,
			// the tables will share the same saved pagination object
			suffix: {
				type: String,
				default: undefined
			}
		}
	});

	const MixinsDeclaration = mixins(Props);

	@Component<PaginatedTable>({
		inheritAttrs: false,
		watch: {
			options(): void {
				this.localStorageUtility.setItem(this.storageKey, this.options);
			}
		}
	})
	export default class PaginatedTable extends MixinsDeclaration {
		localStorageUtility = this.newLocalStorageInstance();

		localOptions = {} as DataOptions;

		newLocalStorageInstance(): LocalStorageUtility {
			if (!this.$vd || !this.$vd.localStorageControl) {
				return new LocalStorageUtility();
			}

			return new LocalStorageUtility(
				this.$vd.localStorageControl.version,
				this.$vd.localStorageControl.expiration,
				this.$vd.localStorageControl.prefix
			);
		}

		get optionsCalc(): DataOptions {
			if (Object.keys(this.localOptions).length) {
				return this.localOptions;
			}

			return this.options;
		}

		set optionsCalc(value: DataOptions) {
			if (Object.keys(this.localOptions).length) {
				this.localOptions = {} as DataOptions;
			}
		}

		get storageKey(): string {
			const PREFIX = 'pagination';

			if (this.suffix) {
				return `${PREFIX}-${this.suffix}`;
			}

			return PREFIX;
		}

		created() {
			this.localOptions = this.localStorageUtility.getItem(this.storageKey) || {} as DataOptions;
		}
	}
</script>
