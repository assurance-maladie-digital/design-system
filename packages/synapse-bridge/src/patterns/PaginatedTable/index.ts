import PaginatedTable from './PaginatedTable.vue';

export default PaginatedTable;
