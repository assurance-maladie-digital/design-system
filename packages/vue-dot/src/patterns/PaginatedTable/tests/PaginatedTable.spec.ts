import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import PaginatedTable from '../';

let wrapper: Wrapper<Vue>;

describe('PaginatedTable', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(PaginatedTable, {
			propsData: {
				options: {},
				serverItemsLength: 0
			}
		});

		expect(wrapper).toMatchSnapshot();
	});
});
