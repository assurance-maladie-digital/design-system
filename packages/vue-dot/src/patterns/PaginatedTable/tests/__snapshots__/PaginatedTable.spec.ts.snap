// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedTable renders correctly 1`] = `<vdatatable-stub mobilebreakpoint="600" items="" options="[object Object]" customsort="[Function]" page="1" itemsperpage="10" groupby="" groupdesc="" customgroup="[Function]" locale="en-US" customfilter="[Function]" serveritemslength="0" itemkey="id" value="" expanded="" noresultstext="$vuetify.dataIterator.noResultsText" nodatatext="$vuetify.noDataText" loadingtext="$vuetify.dataIterator.loadingText" selectablekey="isSelectable" loaderheight="4" headers="" expandicon="$expand" filtermode="intersection" itemclass="" itemstyle=""> </vdatatable-stub>`;
